src/Microsoft.Health.Fhir.Core/Features/Operations/Reindex/ReindexJobTask.cs

```diff
@@ -126,96 +124,172 @@ public async Task ExecuteAsync(ReindexJobRecord reindexJobRecord, WeakETag weakE
         {
             try
             {
-                query.Status = OperationStatus.Running;
-                query.LastModified = DateTimeOffset.UtcNow;
+                SearchResult results;
 
-                // Query first batch of resources
-                var results = await ExecuteReindexQueryAsync(query, false, cancellationToken);
-
-                // if continuation token then update next query
-                if (!string.IsNullOrEmpty(results.ContinuationToken))
+                lock (jobLock)
                 {
-                    var nextQuery = new ReindexJobQueryStatus(results.ContinuationToken);
-                    nextQuery.LastModified = DateTimeOffset.UtcNow;
-                    nextQuery.Status = OperationStatus.Queued;
-                    _reindexJobRecord.QueryList.Add(nextQuery);
-                }
+                    query.Status = OperationStatus.Running;
+                    query.LastModified = DateTimeOffset.UtcNow;
 
-                await UpdateJobAsync(cancellationToken);
+                    // Query first batch of resources
+                    results = ExecuteReindexQueryAsync(query, false, cancellationToken).Result;
 
-                // TODO: Release lock on job document so another thread may pick up the next query.
+                    // if continuation token then update next query
+                    if (!string.IsNullOrEmpty(results.ContinuationToken))
+                    {
+                        var nextQuery = new ReindexJobQueryStatus(results.ContinuationToken);
+                        nextQuery.LastModified = DateTimeOffset.UtcNow;
+                        nextQuery.Status = OperationStatus.Queued;
+                        _reindexJobRecord.QueryList.Add(nextQuery);
+                    }
 
+                    UpdateJobAsync(cancellationToken).Wait();
+                }
+
+                _logger.LogInformation($"Reindex job current thread: {Thread.CurrentThread.ManagedThreadId}");
                 await _reindexUtilities.ProcessSearchResultsAsync(results, _reindexJobRecord.Hash, cancellationToken);
 
-                // TODO: reaquire document lock and update _etag
-                query.Status = OperationStatus.Completed;
-                await UpdateJobAsync(cancellationToken);
+                if (!cancellationToken.IsCancellationRequested)
+                {
+                    lock (jobLock)
+                    {
+                        query.Status = OperationStatus.Completed;
+                        UpdateJobAsync(cancellationToken).Wait();
+                    }
+                }
 
-                await CheckJobCompletionStatus(cancellationToken);
+                return query;
             }
             catch (Exception ex)
             {
-                query.Error = ex.Message;
-
-                query.FailureCount++;
+                lock (jobLock)
+                {
+                    query.Error = ex.Message;
+                    query.FailureCount++;
```

Comment:
We seem to be double counting failures here. If `ExecuteReindexQueryAsync` throws and exception and fails, we increment the failure count there. And then when the exception is caught here, we will increment `FailureCount` again.

Reply:
Yes, removed one of the increments.